Flatten exercises (#27)
eficode-academy/rf-advanced-web-katas: exercises/03-xpath-intro.md

- Implement the keyword to open the form, and make it a part of the test setup.
  - Test should pass with or without the dropdown initially opened.
- **Optional:** Investigate what happens if you don't close the dropdown beforehand.
What errors do you see?

### Step-by-step

Implement the logic to close the dropdown if it's open.

- Using your browser's console, look for a suitable locator to identify an opened dropdown.
- Create a variable in Variables table, called `OPENED DROPDOWN` and give it the value of your locator.
- Add a new keyword called `Close Dropdown If Opened`.
   - Check if the dropdown is opened with a library-specific keyword.
   - Catch the result of that keyword into a local variable, called `${element visible}`.
   - You can use a keyword from the BuiltIn library to click the dropdown if `${element visible}` is true.

### Useful keywords
<details>
  <summary>SeleniumLibrary</summary>

- Use BuiltIn library's `Run Keyword And Return Status` and SeleniumLibrary's `Page Should Contain Element` to check if the dropdown is opened.

</details> <!-- SeleniumLibrary -->

<details>
  <summary>Browser</summary>

> - Use `Get Element State` with `visible` as the state and store it in a variable.

Reviewer: `Get Element State` is actually deprecated, but the whole exercise needs a little work to address that. I'll make a note of that and let's handle this in another PR.

> `dropdown-menu` is visible in the page, after checking that the page is fully loaded, to avoid
> creating race conditions. Often with XPaths, there is more than "one true answer".

</details> <!-- Constructing the xpath. Here's a more detailed explanation. -->
<br/>

Since our locator is pretty generic, we add it to the `Variables` table in our resource file.
Following Robot Framework's best practices, we give our
variable an UPPER CASE name. As we write more code, we can add new XPaths into the table of `Variables`, every time giving them meaningful names.

> We check if the dropdown is open, or not, and store the result in a local variable. We can then use `Run Keyword If` from BuiltIn library, to execute a keyword based on the value of `${element visible}`.

Reviewer: ```suggestion
We check if the dropdown is open, or not, and store the result in a local variable.
We can the use Robot Framework's native inline `IF` to execute a keyword based on the value of `${element visible}`.
```

---

### Implement the keyword to open the form, and make it a part of the test setup.

- Using your browser's console, look for a suitable locator for the button that shows the form.
- Create a variable in Variables table, called `SHOW FORM BUTTON`, and give it the value of your locator.
- Implement the keyword called `Show Form` which clicks that element.
- Add `Test Setup` to your `Settings` table and find a way to call both `Close Dropdown If Opened` and `Show Form`.

> Now we're able to close the dropdown if it's opened but we still need to show our form.

Reviewer: ```suggestion
Now we're able to close the dropdown if it's opened, but we still need to show our form.
```
